docs/gen_cmakelists.man
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GEN_CMAKELISTS 1"
.TH GEN_CMAKELISTS 1 "2022-09-03" "perl v5.26.3" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
gen_cmakelists.pl \- Generate CMakeLists Tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
gen_cmakelists.pl [\fI\s-1OPTIONS\s0\fR] [\-\-] <\fIprofile\fR> <\fIcmakelists\fR>
.PP
.Vb 3
\& OPTIONS:
\&    \-\-config_path <config_path>        set a path of config
\&    \-\-next_minor                       automatic increase minor
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.SS "config"
.IX Subsection "config"
Default path of config is \f(CW\*(C`.gcmake\*(C'\fR, and the config file is \f(CW\*(C`.gcmake/config\*(C'\fR. 
You can define variable in config file and using it in \fIprofile\fR.
Define one variable on each line in config file, and the format is \f(CW\*(C`variable=value\*(C'\fR, 
so after that you can type \f(CW\*(C`@@{variable}@@\*(C'\fR to refer to \f(CW\*(C`variable\*(C'\fR in \fIprofile\fR.
.PP
Support some buildin variable of be called \f(CW\*(C`BUILDIN_VAR\*(C'\fR in config file. The \f(CW\*(C`BUILDIN_VAR\*(C'\fR can be used in \f(CW\*(C`value\*(C'\fR, 
such as has a \f(CW\*(C`BUILDIN_VAR\*(C'\fR of called \f(CW\*(C`VERSION\*(C'\fR, 
then you can define \f(CW\*(C`variable=$VERSION\*(C'\fR in config file and using it in \fIprofile\fR, 
and the value of \f(CW\*(C`variable\*(C'\fR is dynamic that be determined by \f(CW$VERSION\fR.
.PP
.Vb 2
\&    C<BUILDIN_VAR>:
\&        VERSION: major.minor.patch, such as 1.0.0
.Ve
.PP
Such as: have a config as follow:
    STATIC_VERSION=1.2.1
    DYNAMIC_VERSION=$VERSION
.PP
and have content in \fIprofile\fR as follow:
    project(cjson \s-1VERSION\s0 @@{\s-1STATIC_VERSION\s0}@@ \s-1LANGUAGES C CXX\s0)
    project(cjson \s-1VERSION\s0 @@{\s-1DYNAMIC_VERSION\s0}@@ \s-1LANGUAGES C CXX\s0)
.PP
Assume to value of \f(CW$VERSION\fR is \f(CW2.1.1\fR, then content of \fIcmakelists\fR as follow after execused:
    project(cjson \s-1VERSION 1.2.1 LANGUAGES C CXX\s0)
    project(cjson \s-1VERSION 2.1.1 LANGUAGES C CXX\s0)
.SH "EXAMPLES"
.IX Header "EXAMPLES"
\&\f(CW\*(C` ./tools/gen_cmakelists.pl \-\-next_minor \-\-  CMakeLists.txt.in CMakeLists.txt \*(C'\fR
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.SH "AUTHOR"
.IX Header "AUTHOR"
.IP "\(bu" 4
Name: \s-1XXX\s0 (Homqyy)
.IP "\(bu" 4
Email: yilupiaoxuewhq@163.com
.IP "\(bu" 4
Blog: https://www.homqyy.cn
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2022 Homqyy


gen_cmakelists.pl
#!/usr/bin/perl
#

use Getopt::Long;

############################################# Function

use constant {
    KEY    => 0,
    VALUE  => 1,
};

use constant {
    MAJOR_VER   => 0,
    MINOR_VER   => 1,
    PATCH_VER   => 2,
};

sub init_vars
{
    my $gcmake = shift;
    my $config_path = $gcmake->{args}->{config_path};

    my $version_file = "$config_path/.version";
    my $next_ver = "1.1.0";
    my %vars = (
        '$VERSION' => "1.0.0",
    );

    if (-e $version_file)
    {
        open my $version_fd, "<", $version_file
            or die "don't open file $version_file: $!";

        local $/;
        my @ver;
        while (<$version_fd>)
        {
            @ver = (/^(\d++).(\d++).(\d++)\n?$/s);

            die "invalid version in $version_file" unless @ver;

            printf "# current version: %s, %s, %s\n",
                $ver[MAJOR_VER], $ver[MINOR_VER], $ver[PATCH_VER];

        }

        close $version_fd;

        $vars{'$VERSION'} = "$ver[MAJOR_VER].$ver[MINOR_VER].$ver[PATCH_VER]";

        $ver[MINOR_VER] += 1;

        $next_ver = "$ver[MAJOR_VER].$ver[MINOR_VER].$ver[PATCH_VER]"
    }

    $gcmake->{next_version} = $next_ver;

    return \%vars;
}

sub next_version
{
    my $gcmake = shift;
    my $config_path = $gcmake->{args}->{config_path};

    my $version_file = "$config_path/.version";

    print "# next version: $gcmake->{next_version}\n";

    open my $version_fd, "+>", $version_file
        or die "don't open file $version_file: $!";

    print {$version_fd} $gcmake->{next_version};

    close $version_fd;
}

sub parse_config
{
    $gcmake = shift;
    my $config_path = $gcmake->{args}->{config_path};

    my $config_file = "$config_path/config";

    my $conf = {};

    my $vars = &init_vars($gcmake, $config_path);
    gcmake->{vars} = $vars;

    open my $config_fd, "<", $config_file
        or die "don't open file $config_file: $!";

    my $line_n = 0;
    while (<$config_fd>)
    {
        $line_n++;

        chomp;

        next if /^\s*#/;
        next if /^$/;

        my @cl = (/^([^=]++)=([^ ]++)$/);

        printf "# parse: %s, %s\n", $cl[KEY], $cl[VALUE];

        if ($cl[VALUE] =~ '^\$')
        {
            return undef, "invalid variable '$cl[VALUE]' in $config_file: $line_n"
                unless $vars->{$cl[VALUE]};
            
            $cl[VALUE] = $vars->{$cl[VALUE]};
        }

        $conf->{$cl[KEY]} = $cl[VALUE];

        print "# insert: $cl[KEY] => $cl[VALUE]\n";
    }

    close $config_fd;

    print "# parse done\n";
    return $conf, undef;
}

sub get_options
{
    my %args;

    my $config_path   = ".gcmake";
    my $next_minor    = '';
    my $next_major    = '';

    GetOptions (
        "config_path=s" => \$config_path,
        "next_major" => \$next_major,
        "next_minor" => \$next_minor,
    );

    $args{config_path} = $config_path;
    $args{next_minor} = $next_minor;
    $args{next_major} = $next_major;

    return \%args;
}

############################################# Main

my %gcmake;

$gcmake{args} = &get_options;

my ($profile, $cmakelists) = @ARGV;

my $conf, $err;

if (-d $gcmake{args}->{config_path})
{
    ($conf, $err) = &parse_config(\%gcmake);

    if ($err)
    {
        print "$err\n";
        exit 1;
    }
}
else
{
    mkdir $gcmake{args}->{config_path}
}

open my $profile_fd, "<", $profile
    or die "don't open file $profile: $!";

open my $cmakelists_fd, "+>", $cmakelists
    or die "don't open file $cmakelists: $!";

while (<$profile_fd>)
{
    foreach my $key (keys %{$conf})
    {
        s/@@\{$key\}@@/$conf->{$key}/g;
    }

    print {$cmakelists_fd} $_;
}

&next_version($gcmake) if $gcmake{args}->{next_minor};

close $profile_fd;
close $cmakelists_fd;

__END__

=encoding utf8

=head1 NAME

gen_cmakelists.pl - Generate CMakeLists Tool

=head1 SYNOPSIS

gen_cmakelists.pl [I<OPTIONS>] [--] <I<profile>> <I<cmakelists>>

 OPTIONS:
    --config_path <config_path>        set a path of config
    --next_minor                       automatic increase minor

=head1 DESCRIPTION

=head2 config

Default path of config is C<.gcmake>, and the config file is C<.gcmake/config>. 
You can define variable in config file and using it in I<profile>.
Define one variable on each line in config file, and the format is C<variable=value>, 
so after that you can type C<@@{variable}@@> to refer to C<variable> in I<profile>.

Support some buildin variable of be called C<BUILDIN_VAR> in config file. The C<BUILDIN_VAR> can be used in C<value>, 
such as has a C<BUILDIN_VAR> of called C<VERSION>, 
then you can define C<variable=$VERSION> in config file and using it in I<profile>, 
and the value of C<variable> is dynamic what be determined by C<$VERSION>.

    C<BUILDIN_VAR>:
        VERSION: major.minor.patch, such as 1.0.0

Such as: have a config as follow:
    STATIC_VERSION=1.2.1
    DYNAMIC_VERSION=$VERSION

and have content in I<profile> as follow:
    project(cjson VERSION @@{STATIC_VERSION}@@ LANGUAGES C CXX)
    project(cjson VERSION @@{DYNAMIC_VERSION}@@ LANGUAGES C CXX)

Assume to value of C<$VERSION> is C<2.1.1>, then content of I<cmakelists> as follow after execused:
    project(cjson VERSION 1.2.1 LANGUAGES C CXX)
    project(cjson VERSION 2.1.1 LANGUAGES C CXX)

=head1 EXAMPLES

C< ./tools/gen_cmakelists.pl --next_minor --  CMakeLists.txt.in CMakeLists.txt >

=head1 SEE ALSO

=head1 AUTHOR

=over 4

=item * Name: 王鸿奇 (Homqyy)

=item * Email: yilupiaoxuewhq@163.com

=item * Blog: https://www.homqyy.cn

=back

=head1 COPYRIGHT

Copyright (c) 2022 Homqyy

=cut